### PR TITLE
GH-3957 split NOT IN and NOT EXISTS into separate tokens

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #3957  

Briefly describe the changes proposed in this PR:

- in sparql JavaCC grammar, split `NOT_IN` and `NOT_EXISTS` into separate tokens to allow arbitrary whitespace between
- regenerated SPARQL parser from grammar using JavaCC 7.0.4
- added two regression tests

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/ParseException.java [hunk 1]
```diff
@@ -1,12 +1,5 @@
...
- * All rights reserved. This program and the accompanying materials
- * are made available under the terms of the Eclipse Distribution License v1.0
- * which accompanies this distribution, and is available at
- * http://www.eclipse.org/org/documents/edl-v10.php.
- *******************************************************************************/
-/* Generated By:JavaCC: Do not edit this line. ParseException.java Version 5.0 */
-/* JavaCCOptions:KEEP_LINE_COL=null */
+/* Generated By:JavaCC: Do not edit this line. ParseException.java Version 7.0 */
+/* JavaCCOptions:KEEP_LINE_COLUMN=true */
 package org.eclipse.rdf4j.query.parser.sparql.ast;
```
Comment (jeenbroekstra) [inline core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/ParseException.java:3]: Fwiw I'm not particulary fussed about having the copyright headers on these generated files.

**Comment by hmottestad** [comment]
I remember I split up one of the very long methods somewhere in the generated code to help the compiler. I need to check if that is still an issue now.

**Comment by hmottestad** [comment]
@jeenbroekstra do we have some documentation for how to run JavaCC? I could never figure it out.

**Comment by jeenbroekstra** [comment]
It's actually something I keep having to find out myself. I'll add a readme.

**Comment by hmottestad** [comment]
## Before

## After
```
Benchmark                                                     Mode  Cnt      Score     Error  Units
QueryBenchmark.complexQuery                                   avgt    5      4.450 ±   0.032  ms/op
QueryBenchmark.different_datasets_with_similar_distributions  avgt    5      2.332 ±   0.041  ms/op
QueryBenchmark.groupByQuery                                   avgt    5      1.106 ±   0.032  ms/op
QueryBenchmark.long_chain                                     avgt    5    701.260 ±   3.767  ms/op
QueryBenchmark.lots_of_optional                               avgt    5    193.124 ±   0.578  ms/op
QueryBenchmark.minus                                          avgt    5  15395.243 ± 173.074  ms/op
QueryBenchmark.nested_optionals                               avgt    5    204.214 ±   1.559  ms/op
QueryBenchmark.pathExpressionQuery1                           avgt    5     17.498 ±   0.045  ms/op
QueryBenchmark.pathExpressionQuery2                           avgt    5      1.333 ±   0.072  ms/op
QueryBenchmark.query_distinct_predicates                      avgt    5     87.120 ±   0.620  ms/op
QueryBenchmark.simple_filter_not                              avgt    5      9.477 ±   0.192  ms/op
```

**Comment by hmottestad** [comment]
It's about 3x slower. I think it's probably the JVM compiling issue we've run into previously.

**Comment by hmottestad** [comment]
## After splitting up the big SyntaxTreeBuilderTokenManager#jjMoveNfa_0 method

```
Benchmark                                        Mode  Cnt  Score   Error  Units
SPARQLParseBenchmark.complexPathExpressionQuery  avgt    5  8.564 ± 0.857  ms/op
SPARQLParseBenchmark.selectValuesQuery           avgt    5  7.515 ± 0.403  ms/op
```

**Comment by jeenbroekstra** [comment]
> ## After splitting up the big SyntaxTreeBuilderTokenManager#jjMoveNfa_0 method
> 
> ```
> Benchmark                                        Mode  Cnt  Score   Error  Units
> SPARQLParseBenchmark.complexPathExpressionQuery  avgt    5  8.564 ± 0.857  ms/op
> SPARQLParseBenchmark.selectValuesQuery           avgt    5  7.515 ± 0.403  ms/op
> ```

We should do several things:

- create a patch file with this specific fix, so that we can reapply on the generated code after rerunning javacc. I suggest separate patch files for each specific fix.
- add info in a readme file on the how/why of the fix. This is necessary to help fix if the patch file no longer applies without problems (e.g when we bump javacc versions).

**Comment by hmottestad** [comment]
We have an existing patch file for this actually.

Can I propose we create some sort of script that runs JavaCC and applies all the patch files?

Then we could have a nice folder structure:
```
javaCC-sparql-generator
    patches
        large-method-jit
            README.md
            fix.patch
    script.sh
```

**Comment by hmottestad** [comment]
I've started on an issue for this here: https://github.com/eclipse/rdf4j/issues/3962

In the mean time I would like to merge this PR. I also have a[ fix the the serializable issue](https://github.com/eclipse/rdf4j/pull/3963) with SimpleLiteral, if you could review that for me then I will start on releasing 4.0.2 either tonight or tomorrow.

**Review by hmottestad — APPROVED** [review]

**Comment by jeenbroekstra** [comment]
> We have an existing patch file for this actually.

Yes but none of it works, presumably because the previous version of the parser was generated with javacc 5.0. And because none of it is documented or motivated I have no idea how to update that patch (or even which parts of it are still relevant). That patch file should be deleted/replaced.

> 
> Can I propose we create some sort of script that runs JavaCC and applies all the patch files?

I don't think that's a good idea. I'd rather just document the steps. It's not fully automated, there's some inspection and adjustment involved in some steps. 

> 
> Then we could have a nice folder structure:
> ```
> javaCC-sparql-generator
>     patches
>         large-method-jit
>             README.md
>             fix.patch
>     script.sh
> ```
>

**Pull request merged**